File: Test2/test2.1/test2.1/list.cpp
#include "list.h"

struct Node
{
	int value;
	Node *next;
};

struct List
{
	Node *head = nullptr;
};

List *createList()
{
	return new List;
}

bool isEmpty(List *list)
{
	return list->head == nullptr;
}

int listLength(List *list)
{
	Node *current = list->head;
	int listLength = 0;
	while (current != nullptr)
	{
		listLength++;
		current = current->next;
	}
	return listLength;
}

void addValue(List *list, int value)
{
	Node *current = list->head;
	Node *previous = nullptr;
	while (current != nullptr)
	{
		previous = current;
		current = current->next;
	}
	Node *newNode = new Node {value, current};
	previous == nullptr ? list->head = newNode : previous->next = newNode;
}

List *newList(List *list)
{
	if (isEmpty(list))
	{
		return nullptr;
	}
	List *newList = createList();
	int maxLength = 0;
	List *tempList = createList();
	addValue(tempList, list->head->value);
	Node *current = list->head->next;
	Node *previous = list->head;

	while (true)
	{
		while (current != nullptr && current->value > previous->value)
		{
			addValue(tempList, current->value);
			previous = current;
			current = current->next;
		}

		if (listLength(tempList) > maxLength)
		{
			deleteList(newList);
			List *newList = createList();
			newList->head = tempList->head;
			maxLength = listLength(tempList);
			tempList->head = nullptr;
		}
		else
		{
			deleteList(tempList);
			List *tempList = createList();
		}
		if (current == nullptr)
		{
			break;
		}
		addValue(tempList, current->value);
		previous = current;
		current = current->next;
	}
	deleteList(tempList);
	return newList;
}

int deleteValue(List *list)
{
	int headValue = list->head->value;
	Node *temp = list->head->next;
	delete list->head;
	list->head = temp;
	return headValue;
}

void deleteList(List *list)
{
	while (!isEmpty(list))
	{
		Node *temp = list->head->next;
		delete list->head;
		list->head = temp;
	}
	delete list;
}

File: Semester2/Homeworks/HW4/Task2/Task2/List.cs
using Task2.Exceptions;

namespace Task2
{
    public class List
    {        
        private class Node
        {
            public Node(int value, Node next)
            {
                Value = value;
                Next = next;
            }

            public int Value { get; set; }
            public Node Next { get; set; }
        }

        private Node head;

        public int Size { get; private set; }

        public bool IsEmpty => head == null;

        public virtual void AddValue(int value, int position)
        {
            if (position < 0 || position > Size)
            {
                throw new InvalidPositionException();
            }

            Node previous = null;
            var current = head;
            for (var i = 0; i < position; ++i)
            {
                previous = current;
                current = current.Next;
            }

            if (previous == null)
            {
                head = new Node(value, head);
            }
            else
            {
                previous.Next = new Node(value, current);
            }
            Size++;
        }

        public void DeleteValue(int position)
        {
            if (position < 0 || position > Size - 1)
            {
                throw new InvalidPositionException();
            }

            Node previous = null;
            var current = head;
            for (var i = 0; i < position; ++i)
            {
                previous = current;
                current = current.Next;
            }

            if (previous == null)
            {
                head = current.Next;
            }
            else
            {
                previous.Next = current.Next;
            }
            Size--;
        }

        public int GetValue(int position)
        {
            if (position < 0 || position > Size - 1)
            {
                throw new InvalidPositionException();
            }

            var current = head;
            for (var i = 0; i < position; ++i)
            {
                current = current.Next;
            }
            return current.Value;
        }

        public virtual void SetValue(int value, int position)
        {
            if (position < 0 || position > Size - 1)
            {
                throw new InvalidPositionException();
            }

            var current = head;
            for (var i = 0; i < position; ++i)
            {
                current = current.Next;
            }
            current.Value = value;
        }

        public bool Contains(int value)
        {
            var current = head;
            for (var i = 0; i < Size; ++i)
            {
                if (current.Value == value)
                {
                    return true;
                }
                current = current.Next;
            }
            return false;
        }

        public int IndexOf(int value)
        {
            var current = head;
            for (var i = 0; i < Size; ++i)
            {
                if (current.Value == value)
                {
                    return i;
                }
                current = current.Next;
            }
            return -1;
        }
    }
}